### PR TITLE
[SPARK-7031][ThriftServer]let thrift server take SPARK_DAEMON_MEMORY and SPARK_DAEMON_JAVA_OPTS

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
We should let Thrift Server take these two parameters as it is a daemon. And it is better to read driver-related configs as an app submited by spark-submit.

https://issues.apache.org/jira/browse/SPARK-7031

### CODE REVIEW

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 1]
```diff
@@ -190,6 +190,10 @@
       firstNonEmptyValue(SparkLauncher.DRIVER_EXTRA_CLASSPATH, conf, props) : null;
 
     List<String> cmd = buildJavaCommand(extraClassPath);
+    // Take Thrift Server as daemon
+    if (isThriftServer(mainClass)) {
```
Comment (srowen) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: I'm probably missing basic stuff here but it jumps out at me that you have to special-case thrift server here. Is this really related to spark submit?

Comment (WangTaoTheTonic) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: Yeah. I think so. If we take a look at `start-thriftserver.sh` then will find in `spark-daemon.sh` we use spark-submit to submit the `org.apache.spark.sql.hive.thriftserver.HiveThriftServer2`. 

Then it comes into spark-class, and spark-class use `org.apache.spark.launcher.Main` to resolve args in which it has:
>if (className.equals("org.apache.spark.deploy.SparkSubmit")) {
        builder = new SparkSubmitCommandBuilder(args);
      } else {
        builder = new SparkClassCommandBuilder(className, args);
      }

So I think we start Thrift Server by spark-submit but at the meantime Thrift Server is sort of daemon process and it should take daemon related options too.

**Comment by WangTaoTheTonic** [comment]
BTW I have tested on my cluster with setting 
>export SPARK_DAEMON_MEMORY=1111m
export SPARK_DAEMON_JAVA_OPTS=" -Dx=y "

in spark-env.sh.
Before this patch the jinfo shows:
>
VM Flags:
-Xms512m -Xmx512m -XX:MaxPermSize=128m

After:
>VM Flags:
-Dx=y -Xms1111m -Xmx1111m -XX:MaxPermSize=128m

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java (hunk 1 above)
Comment (vanzin) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: The alternative would be to make this `HiveThriftServer2` go through `spark-class` instead of `spark-submit`, but I think that would be a bigger change. While I'm not in love of the idea of peppering special cases inside  `SparkSubmitCommandBuilder`, one doesn't sound like a terrible idea. If there are more in the future we can look at a better solution.

File: launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java [hunk 2]
```diff
@@ -59,6 +59,8 @@ public void testClusterCmdBuilder() throws Exception {
   @Test
   public void testCliParser() throws Exception {
     List<String> sparkSubmitArgs = Arrays.asList(
+      parser.CLASS,
```
Comment (vanzin) [inline launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java]: Are these test changes needed? The tests have been passing without them.

Comment (WangTaoTheTonic) [inline launcher/src/test/java/org/apache/spark/launcher/SparkSubmitCommandBuilderSuite.java]: No they are not necessary and just left while I fixing the broken tests. I will revert these.

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java (hunk 1 above)
Comment (WangTaoTheTonic) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: Total agree with you.

**Comment by WangTaoTheTonic** [comment]
Jenkins, retest this please.

**Comment by WangTaoTheTonic** [comment]
Jenkins, test this please.

**Comment by WangTaoTheTonic** [comment]
Jenkins, test this please.

**Comment by WangTaoTheTonic** [comment]
@vanzin So though it is not a best idea but ok to go?

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 3]
```diff
@@ -201,8 +205,16 @@
       // - SPARK_DRIVER_MEMORY env variable
       // - SPARK_MEM env variable
       // - default value (512m)
-      String memory = firstNonEmpty(firstNonEmptyValue(SparkLauncher.DRIVER_MEMORY, conf, props),
-        System.getenv("SPARK_DRIVER_MEMORY"), System.getenv("SPARK_MEM"), DEFAULT_MEM);
+      String memory;
+      // Take Thrift Server as daemon
+      if (isThriftServer(mainClass)) {
```
Comment (vanzin) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java]: To avoid the copy & paste, I'd do:

    String tsMemory = isThriftServer(mainClass) ? getenv("SPARK_DAEMON_MEMORY") : null;
    memory = firstNonEmpty(tsMemory, ...);

**Comment by vanzin** [comment]
Just a small nit, but otherwise looks good.

**Comment by WangTaoTheTonic** [comment]
Jenkins, retest this please.

**Comment by WangTaoTheTonic** [comment]
Jenkins, retest this please.

**Comment by WangTaoTheTonic** [comment]
Looks like the failed test case is flacky. Jenkins, retest this please.

**Comment by WangTaoTheTonic** [comment]
Another different case failed. Jenkins, retest this please.
